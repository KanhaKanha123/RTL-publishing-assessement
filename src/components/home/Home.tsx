import ArticlesListHorizontal from "../articles/ArticlesListHorizontal";
import ArticlesListVerticle from "../articles/ArticlesListVerticle";
import useGetData from '../../hooks/useGetData';
import Banner from "../banner/Banner";
import ErrorBoundary from "../../error-boundries/ErrorBoundry";
import { useRef } from "react";

const Home = () => {
    let callOnce = useRef<Boolean>(false);
    const {state:{articleData, error, loading},setApiUrl} = useGetData();

    const { bundelItems } = articleData;

    if (!callOnce.current) {
        setApiUrl('bundle-api.json');
    }
    
    callOnce.current = true;

    return (<>
        {loading && <span>Loading.......</span>}

        {articleData && (<div data-testid="home-page" className="home">
            <ErrorBoundary>
                <Banner data-testid="banner-component" artcleData={articleData}></Banner>
            </ErrorBoundary>

            {bundelItems?.length > 0 && <ErrorBoundary>
                <ArticlesListHorizontal data-testid="articlesListHorizontal-component" bundles={bundelItems.slice(0, 3)}></ArticlesListHorizontal>
            </ErrorBoundary>}

            {bundelItems?.length > 0 && <ErrorBoundary>
                <ArticlesListVerticle data-testid="articlesListVerticle-component" bundles={bundelItems.slice(3)}></ArticlesListVerticle>
            </ErrorBoundary>}
        </div>)}

        {error !== '' && <span>{error}</span>}
    </>
    )
}

export default Home;