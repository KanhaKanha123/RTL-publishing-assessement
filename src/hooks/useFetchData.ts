import { useEffect, useReducer, useRef, useState } from 'react';
import { formatData } from '../helpers/formatData';
import { InitialState, articleReducer } from '../store/reducers/articleReducer';
import { getFetchApi } from './services/fetchData';

//custom hook to manage api call to get data
const useFetchData = () => {

    const [state, dispatch] = useReducer(articleReducer, InitialState);
    const [apiUrl, setApiUrl] = useState("bundle-api.json");

    let firstUpdate = useRef<Boolean>(false);

    const getData = async (apiUrl: string) => {
        try {
            //with axios test runner was having some issues so i use fetch here
            getFetchApi(apiUrl).then(res => res.json()).then(res => {
                dispatch({ type: "DATA_SUCCESS", payload: formatData(res) });
            }).catch(err => {
                dispatch({ type: "DATA_ERROR", payload: err });
            });
        } catch (err) {
            console.error(err);
        }

    };

    useEffect(() => {
        try {
            //react 18 re mount the component, so to call only once maintating by useref
            if (!firstUpdate.current) {
                getData(apiUrl);
            }
            firstUpdate.current = true;
        } catch (err) {
            console.error(err);
        }
    }, [apiUrl]);

    return { state, setApiUrl };
}

export default useFetchData;