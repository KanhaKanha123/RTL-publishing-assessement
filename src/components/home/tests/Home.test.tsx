import Home from '../Home';
import { renderHook, waitFor, screen, render } from '@testing-library/react';
import renderer, { act } from 'react-test-renderer';
import useFetchData from '../../../hooks/useFetchData';

describe('Home component component test', () => {

    let fetchMock = global.fetch = jest.fn(() =>
        Promise.resolve({
            json: () => Promise.resolve({ test: 100 }),
        }),
    ) as jest.Mock;

    test("Home component load correctly and called the fetch api", async () => {
        act(() => { render(<Home />) });
        renderHook(async () => { await act(() => { useFetchData() }) });

        waitFor(() => {
            const homePage = screen.getByTestId("home-page");
            expect(fetchMock).toHaveBeenCalled();
            expect(homePage).toBeInTheDocument();
        });


    });

    test("Home component should match snapshot", () => {
        const element = renderer.create(<Home/>);
        expect(element.toJSON()).toMatchSnapshot();
    });
})
