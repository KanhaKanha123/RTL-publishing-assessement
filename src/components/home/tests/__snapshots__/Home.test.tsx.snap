// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component component test Home component should match snapshot 1`] = `
Array [
  <span>
    Loading.......
  </span>,
  <div
    className="home"
    data-testid="home-page"
  >
    <div
      className="banner-wrapper"
      data-testid="banner-wrapper-div"
    >
      <img
        data-testid="banner-image"
      />
      <span
        className="title"
        data-testid="banner-label"
      />
      <h1
        className="bannerTitle"
        data-testid="banner-title"
      />
      <span
        data-testid="banner-description"
      />
    </div>
  </div>,
]
`;
